opm/simulators/linalg/bda/WellContributions.cpp

```diff
@@ -24,79 +24,202 @@
 
 #include <opm/common/OpmLog/OpmLog.hpp>
 #include <opm/common/ErrorMacros.hpp>
+#include <opm/simulators/linalg/bda/openclKernels.hpp>
 #include "opm/simulators/linalg/bda/WellContributions.hpp"
 
 namespace Opm
 {
 
+WellContributions::WellContributions(std::string gpu_mode){
+    if(gpu_mode.compare("cusparse") == 0){
+        cuda_gpu = true;
+    }
+
+    if(gpu_mode.compare("opencl") == 0){
+        opencl_gpu = true;
+    }
+}
 
 void WellContributions::alloc()
 {
     if (num_std_wells > 0) {
 #if HAVE_CUDA
-        allocStandardWells();
-#else
-        OPM_THROW(std::logic_error, "Error cannot allocate on GPU for StandardWells because CUDA was not found by cmake");
+        if(cuda_gpu){
+            allocStandardWells();
+        }
+#endif
+
+#if HAVE_OPENCL
+        if(opencl_gpu){
+            h_Cnnzs_ocl = new double[num_blocks * dim * dim_wells];
+            h_Dnnzs_ocl = new double[num_std_wells * dim_wells * dim_wells];
+            h_Bnnzs_ocl = new double[num_blocks * dim * dim_wells];
+            h_Ccols_ocl = new int[num_blocks];
+            h_Bcols_ocl = new int[num_blocks];
+            val_pointers = new unsigned int[num_std_wells + 1];
+
+            allocated = true;
+        }
+#endif
+
+#if !HAVE_CUDA && !HAVE_OPENCL
+        OPM_THROW(std::logic_error, "Error cannot allocate on GPU because neither CUDA nor OpenCL were found by cmake");
 #endif
     }
 }
 
 WellContributions::~WellContributions()
 {
-#if HAVE_CUDA
-    freeCudaMemory();
-#endif
-    if (h_x) {
-        delete[] h_x;
-        delete[] h_y;
-    }
-
     // delete MultisegmentWellContributions
     for (auto ms : multisegments) {
         delete ms;
     }
     multisegments.clear();
-}
 
+#if HAVE_CUDA
+    if(cuda_gpu){
+        freeCudaMemory(); // should come before 'delete[] h_x'
+    }
+#endif
 
 #if HAVE_OPENCL
-void WellContributions::apply(cl::Buffer& d_x, cl::Buffer& d_y) {
+    if (h_x_ocl) {
+        delete[] h_x_ocl;
+        delete[] h_y_ocl;
+    }
+
+    if(opencl_gpu){
+        if(num_std_wells > 0){
+            delete[] h_Cnnzs_ocl;
+            delete[] h_Dnnzs_ocl;
+            delete[] h_Bnnzs_ocl;
+            delete[] h_Ccols_ocl;
+            delete[] h_Bcols_ocl;
+            delete[] val_pointers;
+        }
+    }
+#endif
+}
+
+#if HAVE_OPENCL
+
+void WellContributions::init(cl::Context *context){
+    d_Cnnzs_ocl = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * num_blocks * dim * dim_wells);
```

Comment:
Sorry, cannot merge, yet. For opm-tests/model1/BASE1_MSW_HFA.DATA this throws an exception:
```
Error: [/home/mblatt/src/dune/opm-gpu/opm-simulators/opm/simulators/linalg/bda/openclSolverBackend.cpp:520] OpenCL Error: clCreateBuffer(-61)
CL_INVALID_BUFFER_SIZE
Program threw an exception: [/home/mblatt/src/dune/opm-gpu/opm-simulators/opm/simulators/linalg/bda/openclSolverBackend.cpp:520] OpenCL Error: clCreateBuffer(-61)
CL_INVALID_BUFFER_SIZE
```
num_blocks is 0 in this case. This model has no Standard wells, hence there should be no allocation here?
Please make sure that the above model works as before. There seems ti be more code that might falsely be executed if there are no standard wells (copyDataToGPU, init,

Reply:
BTW the model runs with master.

Reply:
I'll take a look